cmd/noms/diff/diff.go

```diff
@@ -34,31 +35,33 @@ func shouldDescend(v1, v2 types.Value) bool {
 // Diff writes the diff from |v1| to |v2| to |w|.
 // If |leftRight| is true then the left-right diff is used for ordered sequences - see Diff vs DiffLeftRight in Set and Map.
 func Diff(w io.Writer, v1, v2 types.Value, leftRight bool) error {
-	return diff(diffWriter{w, leftRight}, types.Path{}, nil, v1, v2)
-}
-
-func diff(w diffWriter, p types.Path, key, v1, v2 types.Value) error {
 	if v1.Equals(v2) {
 		return nil
 	}
 
-	if shouldDescend(v1, v2) {
-		switch v1.Type().Kind() {
-		case types.ListKind:
-			return diffLists(w, p, v1.(types.List), v2.(types.List))
-		case types.MapKind:
-			return diffMaps(w, p, v1.(types.Map), v2.(types.Map))
-		case types.SetKind:
-			return diffSets(w, p, v1.(types.Set), v2.(types.Set))
-		case types.StructKind:
-			return diffStructs(w, p, v1.(types.Struct), v2.(types.Struct))
-		default:
-			panic("Unrecognized type in diff function")
-		}
+	dw := diffWriter{w, leftRight}
+
+	if !shouldDescend(v1, v2) {
+		line(dw, DEL, nil, v1)
+		return line(dw, ADD, nil, v2)
 	}
 
-	line(w, DEL, key, v1)
-	return line(w, ADD, key, v2)
+	return diff(dw, types.Path{}, nil, v1, v2)
+}
+
+func diff(w diffWriter, p types.Path, key, v1, v2 types.Value) error {
+	switch v1.Type().Kind() {
```

Comment:
I noticed that `diff` is never called if the values are equal, or if we're not descending. Move checks into Diff.